src/App.tsx

```diff
@@ -1,24 +1,214 @@
+  const [todoFilter, setTodoFilter] = useState<StatusValue>(StatusValue.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [visibleError, setVisibleError] = useState('');
+
+  const fetchTodosFromServer = async () => {
+    try {
+      const fetchedTodos = await getTodos(USER_ID);
+
+      setTodos(fetchedTodos as Todo[]);
+    } catch (error) {
+      setVisibleError('Unable to load a todos');
+    }
+  };
+
+  useEffect(() => {
+    fetchTodosFromServer();
+  }, [tempTodo, loadingTodoIds, visibleError]);
+
+  const completedTodoIds = useMemo(() => (
+    getcompletedTodoIds(todos)
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => (
+    preparedTodos(todos, todoFilter)
+  ), [todos, todoFilter]);
+
+  const removeTodos = useCallback(async (todoIds: number[]) => {
+    try {
+      setLoadingTodoIds(prevIds => [...prevIds, ...todoIds]);
+
+      await Promise.all(
+        todoIds.map(async id => {
+          await removeTodo(id);
+        }),
+      );
+
+      setTodos((previousTodos) => (
+        previousTodos.filter(todo => !todoIds.includes(todo.id))
+      ));
+    } catch (error) {
+      setVisibleError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds([]);
+    }
+  }, []);
+
+  const handleUpdate = useCallback(async (
+    todoIds: number[],
+    newTitle?: string,
+  ) => {
+    try {
+      setLoadingTodoIds((prevIds) => [...prevIds, ...todoIds]);
+
+      await Promise.all(
+        todoIds.map(async id => {
+          const todoToUpdate = todos.find(todo => todo.id === id);
+
+          if (todoToUpdate) {
+            const updatedTodo = await changeTodo(id, {
+              ...todoToUpdate,
+              completed: newTitle
+                ? todoToUpdate.completed
+                : !todoToUpdate.completed,
+              title: newTitle || todoToUpdate.title,
+            });
+
+            setTodos(prevTodos => prevTodos.map(todo => (
+              todo.id === updatedTodo.id
+                ? updatedTodo
+                : todo
+            )));
+          }
+        }),
+      );
+    } catch (error) {
+      setVisibleError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds([]);
+    }
+  }, [todos]);
+
+  const addTodo = useCallback(async (title: string) => {
+    try {
+      const newTodo = {
+        title: title.trim(),
+        completed: false,
+        userId: USER_ID,
+      };
+
+      const tempTodoId = 0;
+
+      setTempTodo({
+        ...newTodo,
+        id: tempTodoId,
+      });
+
+      setLoadingTodoIds([tempTodoId]);
+
+      const addedTodo = await addTodoToServer('/todos', newTodo);
+
+      setTodos((currentTodos) => [...currentTodos, addedTodo]);
+    } catch (error) {
+      setVisibleError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+      setLoadingTodoIds([]);
+      setTodoTitle('');
+    }
+  }, []);
 
-export const App: React.FC = () => {
   if (!USER_ID) {
     return <UserWarning />;
   }
```

Comment:
Alternative renders need to be close to the default render
```suggestion
  if (!USER_ID) {
    return <UserWarning />;
  }
  return (
```

Reply:
@33Orange 
Could you explain what you mean?